File: packages/java/src/main/java/ai/alumnium/client/HttpClient.java
package ai.alumnium.client;

import ai.alumnium.Model;
import ai.alumnium.cli.Cli;
import ai.alumnium.cli.CliException;
import ai.alumnium.result.DoStep;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.IOException;
import java.net.ConnectException;
import java.net.ServerSocket;
import java.net.URI;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.Executors;
import java.util.concurrent.ThreadLocalRandom;
import java.util.concurrent.atomic.AtomicReference;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Thin HTTP client that talks to the alumnium server.
 *
 * <p>Transport details:
 *
 * <ul>
 *   <li>Uses {@link java.net.http.HttpClient} on a virtual-thread executor.
 *   <li>Serialises/deserialises JSON with Jackson.
 * </ul>
 */
public final class HttpClient implements AutoCloseable {

  public static final String DEFAULT_SERVER_HOST = "127.0.0.1";

  /** Same default port as the TypeScript {@code server} CLI command. */
  public static final int DEFAULT_SERVER_PORT = 8013;

  /** Default base URL when none is passed (matches TS server defaults). */
  public static final String DEFAULT_BASE_URL =
      "http://" + DEFAULT_SERVER_HOST + ":" + DEFAULT_SERVER_PORT;

  private static final Logger LOG = LoggerFactory.getLogger(HttpClient.class);
  private static final ObjectMapper MAPPER = new ObjectMapper();
  private static final TypeReference<Map<String, Object>> MAP_OF_OBJECT =
      new TypeReference<Map<String, Object>>() {};

  private final java.net.http.HttpClient http;
  private final String baseUrl;
  private final Model model;
  private final AtomicReference<String> sessionId = new AtomicReference<>();
  private volatile String managedPidName;
  private Thread shutdownHook;

  /**
   * @param url server base URL; null or blank auto-starts a managed server (requires an {@code
   *     alumnium-cli-*} JAR on the classpath); explicit values without a scheme get {@code http://}
   * @param model {@link Model} to be used by the session; {@code null} lets the server pick a
   *     default
   * @param platform driver platform string (e.g. {@code "web-selenium"})
   * @param toolSchemas JSON-serialisable tool schemas (maps or annotated POJOs)
   * @param planner enable the planner agent
   * @param excludeAttributes accessibility-tree attributes to drop before serialising
   */
  public HttpClient(
      String url,
      Model model,
      String platform,
      List<?> toolSchemas,
      boolean planner,
      Set<String> excludeAttributes) {
    this.http =
        java.net.http.HttpClient.newBuilder()
            .executor(Executors.newVirtualThreadPerTaskExecutor())
            .connectTimeout(Duration.ofSeconds(30))
            .build();
    this.baseUrl = resolveUrl(url);

    Map<String, Object> body = new LinkedHashMap<>();
    if (model != null) {
      body.put("provider", model.provider().value());
      body.put("name", model.name());
    }
    body.put("tools", toolSchemas == null ? List.of() : toolSchemas);
    body.put("platform", platform);
    body.put("planner", planner);
    body.put(
        "exclude_attributes",
        excludeAttributes == null ? List.of() : List.copyOf(excludeAttributes));

    JsonNode resp = postJson("/v1/sessions", body, Duration.ofSeconds(30));
    this.model = Model.fromString(resp.path("model").asText(null));
    sessionId.set(resp.path("session_id").asText(null));
  }

  public String baseUrl() {
    return baseUrl;
  }

  public Model model() {
    return model;
  }

  public String sessionId() {
    return sessionId.get();
  }

  @Override
  public void close() {
    quit();
  }

  public void quit() {
    String id = sessionId.getAndSet(null);
    if (id != null) {
      delete("/v1/sessions/" + id, Duration.ofSeconds(30));
    }
    stopManagedServer();
  }

  // region Plans / steps / statements

  public record PlanResult(String explanation, List<String> steps) {}

  public PlanResult planActions(String goal, String accessibilityTree) {
    return planActions(goal, accessibilityTree, "unknown");
  }

  public PlanResult planActions(String goal, String accessibilityTree, String app) {
    JsonNode data =
        postJson(
            "/v1/sessions/" + requireSession() + "/plans",
            Map.of("goal", goal, "accessibility_tree", accessibilityTree, "app", app),
            Duration.ofSeconds(120));
    return new PlanResult(data.path("explanation").asText(""), toStringList(data.path("steps")));
  }

  public record ActionResult(String explanation, List<DoStep> actions) {}

  public ActionResult executeAction(String goal, String step, String accessibilityTree) {
    return executeAction(goal, step, accessibilityTree, "unknown");
  }

  public ActionResult executeAction(
      String goal, String step, String accessibilityTree, String app) {
    JsonNode data =
        postJson(
            "/v1/sessions/" + requireSession() + "/steps",
            Map.of("goal", goal, "step", step, "accessibility_tree", accessibilityTree, "app", app),
            Duration.ofSeconds(120));
    return new ActionResult(
        data.path("explanation").asText(""),
        MAPPER.convertValue(data.path("actions"), new TypeReference<List<DoStep>>() {}));
  }

  public record RetrieveResult(String explanation, Data result) {}

  public RetrieveResult retrieve(
      String statement,
      String accessibilityTree,
      String title,
      String url,
      String screenshot,
      String app) {
    Map<String, Object> body = new LinkedHashMap<>();
    body.put("statement", statement);
    body.put("accessibility_tree", accessibilityTree);
    body.put("title", title);
    body.put("url", url);
    body.put("screenshot", screenshot);
    body.put("app", app == null ? "unknown" : app);
    JsonNode data =
        postJson("/v1/sessions/" + requireSession() + "/statements", body, Duration.ofSeconds(120));
    Object rawResult = MAPPER.convertValue(data.path("result"), Object.class);
    return new RetrieveResult(data.path("explanation").asText(""), Data.looselyTypecast(rawResult));
  }

  public FindElementResult findArea(String description, String accessibilityTree, String app) {
    JsonNode data =
        postJson(
            "/v1/sessions/" + requireSession() + "/areas",
            Map.of(
                "description",
                description,
                "accessibility_tree",
                accessibilityTree,
                "app",
                app == null ? "unknown" : app),
            Duration.ofSeconds(60));
    return new FindElementResult(data.path("id").asInt(), data.path("explanation").asText(""));
  }

  public FindElementResult findElement(String description, String accessibilityTree, String app) {
    JsonNode data =
        postJson(
            "/v1/sessions/" + requireSession() + "/elements",
            Map.of(
                "description",
                description,
                "accessibility_tree",
                accessibilityTree,
                "app",
                app == null ? "unknown" : app),
            Duration.ofSeconds(60));
    JsonNode elements = data.path("elements");
    if (!elements.isArray() || elements.isEmpty()) {
      throw new IllegalStateException(
          "Server returned no elements for description: " + description);
    }
    JsonNode first = elements.get(0);
    return new FindElementResult(first.path("id").asInt(), first.path("explanation").asText(null));
  }

  public String analyzeChanges(
      String beforeTree, String beforeUrl, String afterTree, String afterUrl, String app) {
    Map<String, Object> body =
        Map.of(
            "before", Map.of("accessibility_tree", beforeTree, "url", beforeUrl),
            "after", Map.of("accessibility_tree", afterTree, "url", afterUrl),
            "app", app == null ? "unknown" : app);
    JsonNode data =
        postJson("/v1/sessions/" + requireSession() + "/changes", body, Duration.ofSeconds(120));
    return data.path("result").asText("");
  }

  public Map<String, Object> addExample(String goal, List<String> actions) {
    JsonNode data =
        postJson(
            "/v1/sessions/" + requireSession() + "/examples",
            Map.of("goal", goal, "actions", actions),
            Duration.ofSeconds(30));
    return MAPPER.convertValue(data, MAP_OF_OBJECT);
  }

  public void clearExamples() {
    delete("/v1/sessions/" + requireSession() + "/examples", Duration.ofSeconds(30));
  }

  public Map<String, Object> stats() {
    JsonNode data = getJson("/v1/sessions/" + requireSession() + "/stats", Duration.ofSeconds(30));
    return MAPPER.convertValue(data, MAP_OF_OBJECT);
  }

  public void saveCache() {
    postJson("/v1/sessions/" + requireSession() + "/caches", Map.of(), Duration.ofSeconds(30));
  }

  public void discardCache() {
    delete("/v1/sessions/" + requireSession() + "/caches", Duration.ofSeconds(30));
  }

  /** GET /v1/health. Returns {@code true} if the server reports ready. */
  public boolean health() {
    try {
      HttpResponse<String> resp =
          send(
              HttpRequest.newBuilder(URI.create(baseUrl + "/v1/health"))
                  .timeout(Duration.ofSeconds(5))
                  .GET()
                  .build());
      return resp.statusCode() >= 200 && resp.statusCode() < 300;
    } catch (IOException e) {
      return false;
    }
  }

  private String requireSession() {
    String id = sessionId.get();
    if (id == null) {
      throw new IllegalStateException("Session has been closed");
    }
    return id;
  }

  /**
   * When a URL is provided, normalises and returns it. Otherwise, attempts to start a managed
   * alumnium server on a free port. If no CLI binary is on the classpath, falls back to {@link
   * #DEFAULT_BASE_URL}. Mirrors Python's {@code HttpClient._resolve_url()}.
   */
  private String resolveUrl(String url) {
    if (url != null && !url.isBlank()) {
      return normalizeBaseUrl(url);
    }

    try {
      int port = pickUnusedPort();
      String pidName = buildServerPidName();

      LOG.info("Starting managed alumnium server on port {} (pid name: {})", port, pidName);

      Cli.runServer(
              Map.of(
                  "host",
                  DEFAULT_SERVER_HOST,
                  "port",
                  port,
                  "daemon",
                  true,
                  "daemon_pid",
                  pidName,
                  "daemon_force",
                  true,
                  "daemon_wait",
                  true))
          .check();

      this.managedPidName = pidName;
      this.shutdownHook = new Thread(this::stopManagedServer, "alumnium-server-shutdown");
      Runtime.getRuntime().addShutdownHook(this.shutdownHook);

      String managedUrl = "http://" + DEFAULT_SERVER_HOST + ":" + port;
      LOG.debug("Started managed local server: {} ({})", managedUrl, pidName);
      return managedUrl;
    } catch (IllegalStateException e) {
      LOG.info(
          "CLI binary not available, falling back to default URL {}: {}",
          DEFAULT_BASE_URL,
          e.getMessage());
      return DEFAULT_BASE_URL;
    }
  }

  private void stopManagedServer() {
    String pidName = this.managedPidName;
    if (pidName == null) {
      return;
    }
    this.managedPidName = null;
    try {
      Cli.run("server", "--daemon-kill", "--daemon-pid", pidName, "--daemon-force");
      LOG.debug("Stopped managed local server ({})", pidName);
    } catch (CliException e) {
      LOG.warn("Failed to stop managed server ({}): {}", pidName, e.getMessage());
    }
    if (shutdownHook != null) {
      try {
        Runtime.getRuntime().removeShutdownHook(shutdownHook);
      } catch (IllegalStateException ignored) {
        // JVM already shutting down
      }
      shutdownHook = null;
    }
  }

  private static int pickUnusedPort() {
    try (ServerSocket s = new ServerSocket(0)) {
      return s.getLocalPort();
    } catch (IOException e) {
      throw new RuntimeException("Failed to find a free port", e);
    }
  }

  private static String buildServerPidName() {
    String randomId = Long.toHexString(ThreadLocalRandom.current().nextLong()).substring(0, 7);
    return "server-" + ProcessHandle.current().pid() + "-" + randomId + ".pid";
  }

  private static String normalizeBaseUrl(String url) {
    if (url == null || url.isBlank()) {
      return DEFAULT_BASE_URL;
    }
    String t = url.trim();
    if (t.isEmpty()) {
      return DEFAULT_BASE_URL;
    }
    if (!t.contains("://")) {
      t = "http://" + t;
    }
    return t.replaceAll("/+$", "");
  }

  private JsonNode postJson(String path, Object body, Duration timeout) {
    try {
      String payload = MAPPER.writeValueAsString(body);
      HttpRequest req =
          HttpRequest.newBuilder(URI.create(baseUrl + path))
              .timeout(timeout)
              .header("Content-Type", "application/json")
              .header("Accept", "application/json")
              .POST(HttpRequest.BodyPublishers.ofString(payload))
              .build();
      return parseResponse(send(req));
    } catch (IOException e) {
      throw ioToConnectionError(e);
    }
  }

  private JsonNode getJson(String path, Duration timeout) {
    try {
      HttpRequest req =
          HttpRequest.newBuilder(URI.create(baseUrl + path))
              .timeout(timeout)
              .header("Accept", "application/json")
              .GET()
              .build();
      return parseResponse(send(req));
    } catch (IOException e) {
      throw ioToConnectionError(e);
    }
  }

  private void delete(String path, Duration timeout) {
    try {
      HttpRequest req =
          HttpRequest.newBuilder(URI.create(baseUrl + path)).timeout(timeout).DELETE().build();
      HttpResponse<String> resp = send(req);
      raiseForStatus(resp);
    } catch (IOException e) {
      throw ioToConnectionError(e);
    }
  }

  private HttpResponse<String> send(HttpRequest req) throws IOException {
    try {
      return http.send(req, HttpResponse.BodyHandlers.ofString());
    } catch (InterruptedException e) {
      Thread.currentThread().interrupt();
      throw new IOException("Interrupted during HTTP " + req.method() + " " + req.uri(), e);
    }
  }

  private JsonNode parseResponse(HttpResponse<String> resp) throws IOException {
    raiseForStatus(resp);
    String body = resp.body();
    if (body == null || body.isEmpty()) {
      return MAPPER.nullNode();
    }
    return MAPPER.readTree(body);
  }

  private void raiseForStatus(HttpResponse<String> resp) {
    int code = resp.statusCode();
    if (code >= 200 && code < 300) return;
    String body = resp.body();
    throw new HttpResponseException(
        code,
        "HTTP "
            + code
            + " from "
            + resp.request().method()
            + " "
            + resp.request().uri()
            + ": "
            + (body == null ? "<no body>" : body));
  }

  private RuntimeException ioToConnectionError(IOException e) {
    if (e instanceof ConnectException ce) {
      return new ConnectionClosedException(ce.getMessage(), ce);
    }
    return new ConnectionClosedException(e.getMessage(), e);
  }

  private static List<String> toStringList(JsonNode node) {
    if (node == null || !node.isArray()) return List.of();
    return MAPPER.convertValue(node, new TypeReference<List<String>>() {});
  }

  /** Thrown when the server returned a non-2xx status. */
  public static final class HttpResponseException extends RuntimeException {
    private final int statusCode;

    public HttpResponseException(int statusCode, String message) {
      super(message);
      this.statusCode = statusCode;
    }

    public int statusCode() {
      return statusCode;
    }
  }

  /** Thrown when the HTTP transport could not reach the server. */
  public static final class ConnectionClosedException extends RuntimeException {
    public ConnectionClosedException(String message, Throwable cause) {
      super(message, cause);
    }
  }
}

File: packages/python/src/alumnium/clients/http_client.py
from __future__ import annotations

import atexit
from os import getpid
from secrets import token_hex

from portpicker import pick_unused_port
from requests import ConnectionError, delete, get, post

from ..cli import run_server
from ..logutils import get_logger
from ..models import Model
from ..tools.base_tool import BaseTool
from ..tools.tool_to_schema_converter import convert_tools_to_schemas
from .typecasting import Data, loosely_typecast

logger = get_logger(__name__)

DEFAULT_SERVER_HOST = "127.0.0.1"


class HttpClient:
    def __init__(
        self,
        url: str | None,
        model: Model | None,
        platform: str,
        tools: dict[str, type[BaseTool]],
        planner: bool = True,
        exclude_attributes: set[str] | None = None,
    ):
        self._server_pid: str | None = None
        self.base_url = self._resolve_url(url)
        self.session_id = None

        tool_schemas = convert_tools_to_schemas(tools)

        payload = {
            "tools": tool_schemas,
            "platform": platform,
            "planner": planner,
            "exclude_attributes": list(exclude_attributes or []),
            **(
                {
                    "provider": model.provider.value,
                    "name": model.name,
                }
                if model
                else {}
            ),
        }

        response = post(
            f"{self.base_url}/v1/sessions",
            json=payload,
            timeout=30,
        )
        response.raise_for_status()
        response_data = response.json()

        self.session_id = response_data["session_id"]
        self.model = Model.from_string(response_data["model"])

    def get_health(self) -> dict[str, str]:
        response = get(
            f"{self.base_url}/v1/health",
            timeout=30,
        )
        response.raise_for_status()
        return response.json()

    def quit(self):
        try:
            if self.session_id:
                response = delete(
                    f"{self.base_url}/v1/sessions/{self.session_id}",
                    timeout=30,
                )
                response.raise_for_status()
                self.session_id = None
        except ConnectionError:
            if not self._server_pid:
                raise
            logger.debug("Skipping session cleanup: managed server already stopped")
        finally:
            self._stop_server()

    def plan_actions(self, goal: str, accessibility_tree: str, app: str = "unknown") -> tuple[str, list[str]]:
        """
        Plan actions to achieve a goal.
        Returns:
            A tuple of (explanation, steps).
        """
        response = post(
            f"{self.base_url}/v1/sessions/{self.session_id}/plans",
            json={"goal": goal, "accessibility_tree": accessibility_tree, "app": app},
            timeout=120,
        )
        response.raise_for_status()
        response_data = response.json()
        return (response_data["explanation"], response_data["steps"])

    def add_example(self, goal: str, actions: list[str]):
        response = post(
            f"{self.base_url}/v1/sessions/{self.session_id}/examples",
            json={"goal": goal, "actions": actions},
            timeout=30,
        )
        response.raise_for_status()
        return response.json()

    def clear_examples(self):
        response = delete(
            f"{self.base_url}/v1/sessions/{self.session_id}/examples",
            timeout=30,
        )
        response.raise_for_status()

    def execute_action(
        self, goal: str, step: str, accessibility_tree: str, app: str = "unknown"
    ) -> tuple[str, list[dict]]:
        response = post(
            f"{self.base_url}/v1/sessions/{self.session_id}/steps",
            json={"goal": goal, "step": step, "accessibility_tree": accessibility_tree, "app": app},
            timeout=120,
        )
        response.raise_for_status()
        data = response.json()
        return data["explanation"], data["actions"]

    def retrieve(
        self,
        statement: str,
        accessibility_tree: str,
        title: str,
        url: str,
        screenshot: str | None,
        app: str = "unknown",
    ) -> tuple[str, Data]:
        response = post(
            f"{self.base_url}/v1/sessions/{self.session_id}/statements",
            json={
                "statement": statement,
                "accessibility_tree": accessibility_tree,
                "title": title,
                "url": url,
                "screenshot": screenshot if screenshot else None,
                "app": app,
            },
            timeout=120,
        )
        response.raise_for_status()
        data = response.json()
        return data["explanation"], loosely_typecast(data["result"])

    def find_area(self, description: str, accessibility_tree: str, app: str = "unknown"):
        response = post(
            f"{self.base_url}/v1/sessions/{self.session_id}/areas",
            json={"description": description, "accessibility_tree": accessibility_tree, "app": app},
            timeout=60,
        )
        response.raise_for_status()
        data = response.json()
        return {"id": data["id"], "explanation": data["explanation"]}

    def find_element(self, description: str, accessibility_tree: str, app: str = "unknown") -> dict:
        response = post(
            f"{self.base_url}/v1/sessions/{self.session_id}/elements",
            json={"description": description, "accessibility_tree": accessibility_tree, "app": app},
            timeout=60,
        )
        response.raise_for_status()
        return response.json()["elements"][0]

    def analyze_changes(
        self,
        before_accessibility_tree: str,
        before_url: str,
        after_accessibility_tree: str,
        after_url: str,
        app: str = "unknown",
    ) -> str:
        response = post(
            f"{self.base_url}/v1/sessions/{self.session_id}/changes",
            json={
                "before": {
                    "accessibility_tree": before_accessibility_tree,
                    "url": before_url,
                },
                "after": {
                    "accessibility_tree": after_accessibility_tree,
                    "url": after_url,
                },
                "app": app,
            },
            timeout=120,
        )
        response.raise_for_status()
        return response.json()["result"]

    def save_cache(self):
        response = post(
            f"{self.base_url}/v1/sessions/{self.session_id}/caches",
            timeout=30,
        )
        response.raise_for_status()

    def discard_cache(self):
        response = delete(
            f"{self.base_url}/v1/sessions/{self.session_id}/caches",
            timeout=30,
        )
        response.raise_for_status()

    @property
    def stats(self):
        response = get(
            f"{self.base_url}/v1/sessions/{self.session_id}/stats",
            timeout=30,
        )
        response.raise_for_status()
        return response.json()

    def _resolve_url(self, url_option: str | None) -> str:
        if url_option:
            return url_option.rstrip("/")

        port = pick_unused_port()
        pid_name = self._build_server_pid_name(port)

        run_server(
            host=DEFAULT_SERVER_HOST,
            port=port,
            daemon=True,
            daemon_pid=pid_name,
            daemon_force=True,
            daemon_wait=True,
            check=True,
        )

        # Ensure to stop the server when the program exits
        atexit.register(self._stop_server)

        self._server_pid = pid_name
        managed_url = f"http://{DEFAULT_SERVER_HOST}:{port}"
        logger.debug(f"Started managed local server: {managed_url} ({pid_name})")
        return managed_url

    def _stop_server(self) -> None:
        if not self._server_pid:
            return

        run_server(
            daemon_kill=True,
            daemon_pid=self._server_pid,
            daemon_force=True,
        )
        logger.debug(f"Stopped managed local server ({self._server_pid})")
        self._server_pid = None

    @staticmethod
    def _build_server_pid_name(port: int) -> str:
        random_id = token_hex(4)[:7]
        return f"server-{getpid()}-{random_id}.pid"
